File: zig/src/indicators/core/outputs/levels.zig
const std = @import("std");

/// A single entry of a Levels output, expressed as a value with an optional bar
/// offset and an optional strength.
pub const Level = struct {
    /// The value (e.g. a price level or a multiplier) at this entry.
    value: f64,
    /// The number of bars back from the Levels' time (0 = current bar).
    offset: i32 = 0,
    /// Optional significance measure (higher = more significant); NaN when not applicable.
    strength: f64 = std.math.nan(f64),
};

/// Maximum number of entries in a Levels output.
pub const max_levels = 256;

/// Holds a time stamp and a variable-length set of levels.
pub const Levels = struct {
    time: i64,
    levels: [max_levels]Level = undefined,
    levels_len: usize = 0,

    /// Creates a new Levels from a slice of entries.
    pub fn new(time: i64, levels: []const Level) Levels {
        var l = Levels{
            .time = time,
            .levels_len = @min(levels.len, max_levels),
        };
        @memcpy(l.levels[0..l.levels_len], levels[0..l.levels_len]);
        return l;
    }

    /// Creates a new empty Levels with no entries.
    pub fn empty(time: i64) Levels {
        return .{ .time = time };
    }

    /// Indicates whether this Levels has no entries.
    pub fn isEmpty(self: Levels) bool {
        return self.levels_len == 0;
    }

    /// Returns the entries as a slice.
    pub fn levelsSlice(self: *const Levels) []const Level {
        return self.levels[0..self.levels_len];
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

const testing = std.testing;

test "Levels new and slice" {
    const entries = [_]Level{
        .{ .value = 105.5, .offset = 3, .strength = 0.8 },
        .{ .value = 102.0, .offset = 1, .strength = 0.6 },
    };
    const l = Levels.new(42, &entries);
    try testing.expectEqual(@as(i64, 42), l.time);
    try testing.expectEqual(@as(usize, 2), l.levels_len);
    try testing.expectEqual(@as(f64, 105.5), l.levelsSlice()[0].value);
    try testing.expectEqual(@as(i32, 3), l.levelsSlice()[0].offset);
    try testing.expectEqual(@as(f64, 0.8), l.levelsSlice()[0].strength);
    try testing.expect(!l.isEmpty());
}

test "Levels empty" {
    const l = Levels.empty(7);
    try testing.expectEqual(@as(i64, 7), l.time);
    try testing.expect(l.isEmpty());
    try testing.expectEqual(@as(usize, 0), l.levelsSlice().len);
}

test "Level defaults" {
    const lv = Level{ .value = 42.0 };
    try testing.expectEqual(@as(i32, 0), lv.offset);
    try testing.expect(std.math.isNan(lv.strength));
}

File: rs/src/indicators/core/outputs/levels.rs
/// A single entry of a Levels output, expressed as a value with an optional bar
/// offset and an optional strength.
#[derive(Debug, Clone, Copy)]
pub struct Level {
    /// The value (e.g. a price level or a multiplier) at this entry.
    pub value: f64,
    /// The number of bars back from the Levels' time (0 = current bar).
    pub offset: i32,
    /// Optional significance measure (higher = more significant); NaN when not applicable.
    pub strength: f64,
}

impl Level {
    /// Creates a level with the given value, offset and strength.
    pub fn new(value: f64, offset: i32, strength: f64) -> Self {
        Self { value, offset, strength }
    }

    /// Creates a level with the given value, an offset of 0 and a NaN strength.
    pub fn value_only(value: f64) -> Self {
        Self { value, offset: 0, strength: f64::NAN }
    }
}

/// Holds a time stamp and a variable-length set of levels.
#[derive(Debug, Clone)]
pub struct Levels {
    pub time: i64,
    pub levels: Vec<Level>,
}

impl Levels {
    /// Creates a new Levels with the given time and entries.
    pub fn new(time: i64, levels: Vec<Level>) -> Self {
        Self { time, levels }
    }

    /// Creates a new empty Levels with no entries.
    pub fn empty(time: i64) -> Self {
        Self { time, levels: Vec::new() }
    }

    /// Indicates whether this Levels has no entries.
    pub fn is_empty(&self) -> bool {
        self.levels.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_and_fields() {
        let l = Levels::new(42, vec![Level::new(105.5, 3, 0.8), Level::new(102.0, 1, 0.6)]);
        assert_eq!(l.time, 42);
        assert_eq!(l.levels.len(), 2);
        assert_eq!(l.levels[0].value, 105.5);
        assert_eq!(l.levels[0].offset, 3);
        assert_eq!(l.levels[0].strength, 0.8);
        assert!(!l.is_empty());
    }

    #[test]
    fn test_value_only_has_nan_strength() {
        let lv = Level::value_only(42.0);
        assert_eq!(lv.value, 42.0);
        assert_eq!(lv.offset, 0);
        assert!(lv.strength.is_nan());
    }

    #[test]
    fn test_empty() {
        let l = Levels::empty(7);
        assert_eq!(l.time, 7);
        assert!(l.is_empty());
    }
}
